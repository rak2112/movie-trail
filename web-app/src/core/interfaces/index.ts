type MovieType =
  | 'Action'
  | 'Adventure'
  | 'Animation'
  | 'Comedy'
  | 'Crime'
  | 'Documentry'
  | 'Drama'
  | 'Family'
  | 'Fantasy'
  | 'History'
  | 'Horror'
  | 'Music'
  | 'Mystery'
  | 'Romance'
  | 'Science Fiction'
  | 'TV Movie'
  | 'Thriller'
  | 'War'
  | 'Western';

export type UserMovieType= 'FAVORITE' | 'WATCHLIST';

export type UserView = 'Favorites' | 'Watchlist';

export interface Api {
  isFetching?: boolean;
  hasError?: boolean;
  errorMessage?: string;
}; 

export interface FormValues {
  name?: string;
  email: string;
  password: string;
  confirmPassword?: string;
  token?: string;
}

export interface Language {
  id: string;
  name: string;
}

export interface Movie {
  id: number;
  backdrop_path: string;
  genre_ids: number[];
  genres: Genre[];
  homepage: string;
  overview: string;
  poster_path?: string;
  runtime: string;
  release_date: string;
  spoken_languages: Language[];
  status: string;
  title: string;
  type?: string;
  vote_average: string;
};

export interface Movies {
  pageNo: number;
  totalPages: number;
  totalResults: number;
  results: Movie[];
  redirection? : boolean;
};

export interface MoviesData {
  pageNo: number;
  total_results: number;
  total_pages: number;
  results?: Movie[];
};

export interface MoviesRes {
  movies: Movie[]
};

export interface Genre {
  id: number;
  name: MovieType;
};

export interface GenresMap {
  [id: number]: Genre;
};

export interface UserMovie extends Movie {
  _id?: string;
  movieType?: UserMovieType;
}

export interface User {
  email?: string;
  displayName?: string;
  loggedIn?: boolean;
  movies?: UserMovie[];
  resetMessage?: string;
  token?: string;
}

export interface UserMovies {
  [id: string]: UserMovie;
}
export interface UserMovieMap {
  favorites: UserMovies;
  watchlist: UserMovies;
}

export interface Cast {
  character: string;
  id: string;
  name: string;
  profile_path?: string;
}

export interface Crew extends Cast {
  department: string;
  job: string;
}

export interface Poster {
  file_path: string;
}

export interface Persons {
  id?: string;
  cast: Cast[];
  crew: Crew[];
}

export interface Posters {
  id?: string;
  backdrops: Poster[];
  posters: Poster[];
}

export interface Video {
  id: string;
  key: string;
  name: string;
}

export interface Videos {
  id: string;
  results: Video[]
}

export interface MovieDetail {
  isFetching?: boolean;
  hasError?: boolean;
  details: Movie;
  images: Posters;
  persons: Persons;
  videos: Videos
}