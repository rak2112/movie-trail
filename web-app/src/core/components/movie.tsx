import React, { memo } from 'react';
import { Link } from 'react-router-dom';

import UserActions from '../containers/user-actions';
import { Genre, GenresMap, Movie, UserMovieMap, UserView } from '../interfaces';
import { getUuid, paths } from '../utils/util-service';
import { Container, Detail, MovieContainer, NoPoster } from './styles/movie.style';


interface Props {
  genres: GenresMap;
  movie: Movie;
  id?: number;
  profileView?: boolean;
  userView?: UserView;
  userMovies?: UserMovieMap;
  addMovie?: (movie: Movie) => void;
  deleteMovie?: (movie: Movie) => void;
}

export const MovieComponent = memo(({
  genres,
  movie,
  profileView,
  userView
}: Props) => {
  const { backdrop_path, id, genre_ids, genres: detailGrenres, title } = movie;
  return (
    <Container>
      <MovieContainer>
        {
          backdrop_path ?
            <Link to={{pathname: '/movie-details/' + id }}>
              <img src={`${paths.imgPath500}${backdrop_path} `} alt="main image" />
            </Link>
            : <NoPoster>No Image Available</NoPoster>
        }
        <UserActions
          movie={movie}
          userView={userView}
          profileView={profileView}
        />
        <Detail>
          <h3>{title}</h3>
          { (genre_ids) ? 
            genre_ids.map((genre: number) => ( genres && genres[genre] && <span className="genre" key={getUuid()}>{ genres[genre].name }</span>))
            : 
            detailGrenres.map((genre: Genre) =>  <span className="genre" key={getUuid()}>{ genre.name }</span> )
          }
          <p>
          <Link to={{pathname: '/movie-details/'+ id }}>View Details</Link>
          </p>
        </Detail>
      </MovieContainer>
    </Container>
  );
});


