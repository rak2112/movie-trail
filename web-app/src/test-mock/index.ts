import { Genre, Movie, Movies, UserMovieMap } from '../core/interfaces';

import { State } from '../core/reducers';

export const defaultMoviesState = {
  addMovie: jest.fn(),
  deleteMovie: jest.fn(),
  genres:[
    {id:2, name: "Action"} as Genre
  ], 
  loadMovies: jest.fn(),
  movies: {
    totalPages: 1, pageNo: 1, totalResults: 3, 
    results: [{
      id: 2,
      backdrop_path: 'path',
      genre_ids: [1, 2],
      genres: [],
      homepage: 'homepage',
      overview: 'overview',
      runtime: '107',
      release_date: '2-04-19',
      spoken_languages: [],
      status: 'status',
      title: 'title',
      vote_average: 'vote_avg'
    },
    {
      id: 3,
      backdrop_path: 'path',
      genre_ids: [1, 2],
      genres: [],
      homepage: 'homepage',
      overview: 'overview',
      runtime: '107',
      release_date: '2-04-19',
      spoken_languages: [],
      status: 'status',
      title: 'title',
      vote_average: 'vote_avg'
    }]
  } as Movies,
  userMovies: {
    favorites: {
      2: {
        id: 2,
        _id: '23',
        backdrop_path: 'path',
        genre_ids: [1, 2],
        genres: [],
        homepage: 'homepage',
        movieType: 'FAVORITE',
        overview: 'overview',
        runtime: '107',
        release_date: '2-04-19',
        spoken_languages: [],
        status: 'status',
        title: 'title',
        vote_average: 'vote_avg'
      }
    },
    watchlist: {}
  } as UserMovieMap
};

export const defaultState = {
  api: { isFetching: true, hasError: false },
  genres: [{id:2, name: 'Action'} as Genre ],
  movies: null,
  movieDetails: null,
  searchedMovies: null,
  user: null,
  userMovies: {}
} as State;

export const defaultMovies = {
  totalPages: 1, pageNo: 1, totalResults: 3, 
  results: [{
    id: 2,
    backdrop_path: 'path',
    genre_ids: [1, 2],
    genres: [],
    homepage: 'homepage',
    overview: 'overview',
    runtime: '107',
    release_date: '2-04-19',
    spoken_languages: [],
    status: 'status',
    title: 'title',
    vote_average: 'vote_avg'
  },
  {
    id: 3,
    backdrop_path: 'path',
    genre_ids: [1, 2],
    genres: [],
    homepage: 'homepage',
    overview: 'overview',
    runtime: '107',
    release_date: '2-04-19',
    spoken_languages: [],
    status: 'status',
    title: 'title',
    vote_average: 'vote_avg'
  }] as Movie[]
} as Movies;