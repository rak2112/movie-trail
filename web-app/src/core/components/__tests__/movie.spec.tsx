import React from "react";
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';

import { StaticRouter } from 'react-router-dom'
import { create } from 'react-test-renderer';

import { GenresMap, Movie, UserMovieMap } from '../../interfaces';
import { MovieComponent } from '../movie';


import { defaultState } from '../../../test-mock';

// jest.mock('../../utils/util-service');
// const util = require('../../utils/util-service');

const props = {
  addMovie: jest.fn(),
  deleteMovie: jest.fn(),
  genres: {2: {id: 2, name: 'Action'}} as GenresMap,
  loadMovies: jest.fn(),
  movie: {
    id: 2,
    backdrop_path: 'path',
    genre_ids: [1, 2],
    genres: [],
    homepage: 'homepage',
    overview: 'overview',
    runtime: '107',
    release_date: '2-04-19',
    spoken_languages: [],
    status: 'status',
    title: 'title',
    vote_average: 'vote_avg'
  } as Movie,
  userMovies: {
    favorites: {},
    watchlist: {}
  } as UserMovieMap
};

let initialState;
let store: any;
describe(`MovieComponent`, () => {
  

  beforeEach(() => {
    initialState = defaultState;
    store = configureMockStore()(initialState);
  });

  it(`should match the snapshot with given inputs`, () => {
    const context = {};
    const component = create(
      <Provider store={store}>
        <StaticRouter location="someLocation" context={context}>
          <MovieComponent {...props}/>
        </StaticRouter>
      </Provider>
      
    ).toJSON();

    expect(component).toMatchSnapshot();
  });
});