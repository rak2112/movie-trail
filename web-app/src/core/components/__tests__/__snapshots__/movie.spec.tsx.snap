// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieComponent should match the snapshot with given inputs 1`] = `
<div
  className="css-h5tpbd"
>
  <div
    className="css-bxd89m"
  >
    <a
      href="/movie-details/2"
      onClick={[Function]}
    >
      <img
        alt="main image"
        src="https://image.tmdb.org/t/p/w500path "
      />
    </a>
    <div
      className="css-1nkcf6k"
    >
      <div
        className="user-actions"
      >
        <span
          className="rating"
        >
          vote_avg
          <span
            className="glyphicon glyphicon-star"
          />
        </span>
        <span
          className="year"
        >
          <span
            className="glyphicon glyphicon-calendar"
          />
          2019
        </span>
      </div>
    </div>
    <div
      className="css-1w2z22x"
    >
      <h3>
        title
      </h3>
      <span
        className="genre"
      >
        Action
      </span>
      <p>
        <a
          href="/movie-details/2"
          onClick={[Function]}
        >
          View Details
        </a>
      </p>
    </div>
  </div>
</div>
`;
